### PR TITLE
Bring the power of JITMs to Calypso! Part II

### PR DESCRIPTION
**Pull request by withinboredom**
> JITM: (pronounced Jit-em) - Just-in-time message for sending a specified message when a site is in a specific state 

This PR allows jitm dismissals to be persisted on the Jetpack site such that a jitm can be dismissed in both places with one dismissal to prevent ourselves from annoying our users.

# Testing

Run Calypso locally and visit the comments section of a Jetpack site and a free plan. You should see a JITM for Akismet at the top.

### CODE REVIEW

File: client/blocks/jitm/index.jsx [hunk 1]
```diff
@@ -19,25 +22,32 @@ export const JITM = ( {
...
-const mapStateToProps = ( state ) => (
-	{
-		currentSite: getSelectedSite( state ),
-		hasJitm: !! getTopJITM( state ),
-		...getTopJITM( state ),
-	}
-);
+const mapDispatchToProps = dispatch => ( {
+	onDismiss: ( siteId, id, featureClass ) => () =>
+		dismissJetpackJITM( dispatch, siteId, id, featureClass ),
```
Comment (dmsnell) [inline client/blocks/jitm/index.jsx]: this is probably missing a call to `dispatch()`…

**Comment by mattwiebe** [comment]
@withinboredom I'll try to test this later, but:

1. I need some testing directions. I'm aware of JITMs in Jetpack, but what behaviour should I expect? How do I verify this is working as expected? (This is not only because I'm dumb, which I frequently am, but because it helps to document what this should be doing for anyone else, and often has helped me to catch things in the past, too)
2. There's a merge conflict right now

**Comment by withinboredom** [comment]
@mattwiebe added a testing section and rebased.

File: client/blocks/jitm/index.jsx (hunk 1 above)
Comment (withinboredom) [reply client/blocks/jitm/index.jsx]: @dmsnell, what do you mean?

Comment (dmsnell) [reply client/blocks/jitm/index.jsx]: oh, I actually read it wrong, but I'm glad you asked - we definitely don't want to be passing `dispatch` in to an action creator. _this_ place in the code is the appropriate place to dispatch an action - we want the action creators to be able to create their action objects without affecting the system, leaving the actual effects (triggered by `dispatch`) to be left at the edges, which in this case are on a button click.

```js
onDismiss: ( … ) => dispatch( dismissJetpackJITM( siteId, id, featureClass ) );
```

File: client/state/jitm/actions.js [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+export const dismissJetpackJITM = ( dispatch, siteId, id, featureClass ) => {
+	dispatch(
+		{
+			type: JITM_DISMISS,
+			siteId,
+			id,
+			featureClass,
+		}
+	);
+};
```
Comment (dmsnell) [inline client/state/jitm/actions.js]: This is something we would probably be caring more about to make sure we don't dispatching from action creators. Instead we'd much rather have the action creators remain simple functions which return action objects and let someone else handle the dispatch.

File: client/state/jitm/actions.js [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+ * @param {string} messagePath The path of the jitm (ex: "calypso:comments:admin_notices")
+ * @param {object} jitms The objects to display
+ * @return {undefined}
+ */
+export const insertJITM = ( dispatch, siteId, messagePath, jitms ) =>
+	dispatch( {
+		type: JITM_SET,
+		keyedPath: messagePath + siteId,
+		jitms: jitms.map( jitm => ( { ...jitm, lastUpdated: Date.now() } ) ),
+	} );
```
Comment (dmsnell) [inline client/state/jitm/actions.js]: see previous comment

File: client/state/jitm/actions.js [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+ * @param {int} siteId The site identifier
+ * @param {string} messagePath The path of the jitm (ex: "calypso:comments:admin_notices")
+ * @return {undefined}
+ */
+export const clearJITM = ( dispatch, siteId, messagePath ) =>
+	dispatch( {
+		type: JITM_SET,
+		keyedPath: messagePath + siteId,
+		jitms: [],
+	} );
```
Comment (dmsnell) [inline client/state/jitm/actions.js]: see previous previous comment

File: client/state/data-layer/wpcom/sites/jitm/index.js [hunk 5]
```diff
@@ -31,10 +32,20 @@ const process = {
 	lastSite: null,
 };
 
+/**
+ * Existing libraries do not escape decimal encoded entities that php encodes, this handles that.
+ * @param {string} str The string to decode
```
Comment (dmsnell) [inline client/state/data-layer/wpcom/sites/jitm/index.js]: nitpick, but it's nice to have an empty line here between the description and the var types

**Pull request merged**